README.md
# MeerTRAP Trigger Tools #

This repository contains realtime VOEvent triggering software for the [MeerTRAP project](https://www.meertrap.org/) at the MeerKAT telescope. The code is mainly developed for Python 3 (and in particular version 3.7), but Python 2 (e.g. version 2.7) should work fine.

## Author ##

The software is primarily developed and maintained by Fabian Jankowski. For more information, feel free to contact me via: fabian.jankowski at manchester.ac.uk.

## Citation ##

If you make use of the software, please add a link to this repository and cite our upcoming paper. See the CITATION file.

## Installation ##

The easiest and recommended way to install the software is through `pip` directly from the bitbucket repository. For example, to install the master branch of the code, use the following command:

`pip3 install git+https://bitbucket.org/jankowsk/meertrig.git@master`

## Example Usage ##

The `meertrig` software makes it easy to construct and emit a VOEvent packet to a given VOEvent broker. An example is given below.

```python
from meertrig import VOEvent

params = {
    'utc': utc,
    'title': 'Detection of test event',
    'short_name': 'Test event',
    'beam_semi_major': semiMaj,
    'beam_semi_minor': semiMin,
    'beam_rotation_angle': 0.0,
    'tsamp': 0.367,
    'cfreq': 1284.0,
    'bandwidth': 856.0,
    'nchan': 4096,
    'beam': 123,
    'dm': dm,
    'dm_err': dm_err,
    'width': width,
    'snr': snr,
    'flux': flux,
    'ra': coord.ra.deg,
    'dec': coord.dec.deg,
    'gl': coord.galactic.l.deg,
    'gb': coord.galactic.b.deg,
    'name': name,
    'importance': importance,
    'internal': 1,
    'open_alert': 0,
    'test': 1,
    'product_id': product_id
}

v = VOEvent(host=hostname, port=port)
ve = v.generate_event(params, True)
v.send_event(ve)
```


CHANGES.md
# Version History #

## HEAD ##

## 0.2 (untagged) ##

* Simplified the import of the VOEvent class and added representation and string functions to it.
* Bumped version number to 0.2.
* Added installation instructions, example usage and improved the python setup script.
* Updated the equation to compute the inferred redshift to a more recent one (Zhang 2018).
* Automatically look up the Milky Way DM contribution for a given event based on the `YMW16` model. The lookup happens fully automatically in the `VOEvent` class.

## 0.1 (2020-03-09) ##

The code was successfully tested in a TUSE to FBFUSE high-level triggering run.

* Use YAML configuration file with default values.
* Convert main part of the code to VOEvent class that has member functions to generate and emit events, among others.
* Added script to emit VOEvents periodically with randomised (but sensible!) parameters. This will be used to test the receiving code.
* Added script to generate VOEvents from the command line.
* Initial docker software configuration, including various VOEvent python tools.


meertrig/gen_random.py
#
#   2020 Fabian Jankowski
#   Generate periodic VOEvents with randomised content.
#

import argparse
from time import sleep

from astropy.coordinates import SkyCoord
from astropy.time import Time
import astropy.units as units
import numpy as np

from meertrig.config_helpers import get_config
from meertrig.voevent import VOEvent

# astropy.units generates members dynamically, pylint therefore fails
# disable the corresponding pylint test for now
# pylint: disable=E1101


def parse_args():
    """
    Parse the commandline arguments.
    """

    parser = argparse.ArgumentParser(
        description="Generate VOEvents with random content."
    )

    parser.add_argument(
        "--period",
        type=float,
        default=3600,
        help="Emit one event every period seconds (default: 3600).",
    )

    args = parser.parse_args()

    return args


def generate_random_event(v, t_defaults, nr):
    """
    Generate a random VOEvent.

    Parameters
    ----------
    v: ~meertrig.VOEvent
        VOEvent instance.
    t_defaults: dict
        Dictionary with default parameters for the VOEvents.
    nr: int
        The number of the event.
    """

    params = t_defaults.copy()

    # generate random parameters
    utc = Time.now().iso
    ra = np.random.uniform(0, 360)
    dec = np.random.uniform(-90, 25)
    beam = np.random.randint(1, 768)
    dm = np.random.uniform(200, 5000)
    width = np.random.uniform(2, 30)
    snr = np.random.uniform(10, 50)
    flux = 0.1 * snr
    name = "TestEvent{0}".format(nr)
    importance = np.random.uniform(0.5, 1)
    product_id = "array_1"

    # parse coordinates
    coord = SkyCoord(ra=ra, dec=dec, unit=(units.degree, units.degree), frame="icrs")

    event_params = {
        "utc": utc,
        "title": "Detection of test event",
        "short_name": "Test event",
        "beam_semi_major": 64.0 / 60.0,
        "beam_semi_minor": 28.8 / 60.0,
        "beam_rotation_angle": 0.0,
        "tsamp": 0.367,
        "cfreq": 1284.0,
        "bandwidth": 856.0,
        "nchan": 4096,
        "beam": beam,
        "dm": dm,
        "dm_err": 1.0,
        "width": width,
        "snr": snr,
        "flux": flux,
        "ra": coord.ra.deg,
        "dec": coord.dec.deg,
        "gl": coord.galactic.l.deg,
        "gb": coord.galactic.b.deg,
        "name": name,
        "importance": importance,
        "internal": 1,
        "open_alert": 0,
        "test": 1,
        "product_id": product_id,
    }

    params.update(event_params)

    vostr = v.generate_event(params, True)

    return vostr


#
# MAIN
#


def main():
    args = parse_args()

    config = get_config("config.yml")
    defaults = get_config("defaults.yml")

    # treat email address
    defaults["contact_email"] = defaults["contact_email"].replace(" AT ", "@")

    v = VOEvent(host=config["broker"]["host"], port=config["broker"]["port"])

    nr = 0

    while True:
        ve = generate_random_event(v, defaults, nr)
        v.send_event(ve)

        nr += 1
        sleep(args.period)

    print("All done.")


if __name__ == "__main__":
    main()
